File: ios/Bonfire/Core/Chat/PersistedChatService.swift
import Foundation
import RxSwift

final class PersistedChatService: ChatService {
    private let chatDatabase: ChatDatabase

    init(chatDatabase: ChatDatabase) {
        self.chatDatabase = chatDatabase
    }

    func chat(channel: Channel) -> Observable<DatabaseResult<Chat>> {
        return chatDatabase.chat(channel)
            .map({.Success($0)})
            .catchError({Observable.just(.Error($0))})
    }

    func sendMessage(message: Message, channel: Channel) {
        chatDatabase.sendMessage(message, channel: channel)
    }
}

File: android/core/src/main/java/com/novoda/bonfire/chat/service/PersistedChatService.java
package com.novoda.bonfire.chat.service;

import com.novoda.bonfire.channel.data.model.Channel;
import com.novoda.bonfire.chat.data.model.Chat;
import com.novoda.bonfire.chat.data.model.Message;
import com.novoda.bonfire.chat.database.ChatDatabase;
import com.novoda.bonfire.database.DatabaseResult;

import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Function;

public class PersistedChatService implements ChatService {

    private final ChatDatabase chatDatabase;

    public PersistedChatService(ChatDatabase chatDatabase) {
        this.chatDatabase = chatDatabase;
    }

    @Override
    public Observable<DatabaseResult<Chat>> getChat(final Channel channel) {
        return chatDatabase.observeChat(channel)
                .map(asDatabaseResult())
                .onErrorReturn(DatabaseResult.<Chat>errorAsDatabaseResult());
    }

    private static Function<Chat, DatabaseResult<Chat>> asDatabaseResult() {
        return new Function<Chat, DatabaseResult<Chat>>() {
            @Override
            public DatabaseResult<Chat> apply(@NonNull Chat chat) throws Exception {
                return new DatabaseResult<Chat>(chat);
            }
        };
    }

    @Override
    public void sendMessage(Channel channel, Message message) {
        chatDatabase.sendMessage(channel, message);
    }

}
